File: agents/browser/src/session.ts
// Session and page view state (docs/contracts/rum.md §1.1).
//
// A session is a visit, not a person: a random id in sessionStorage, expiring after 30 minutes of inactivity
// and capped at 4 hours. It is never derived from anything about the visitor, is per tab, and disappears when
// the tab does — so it cannot be used to recognise someone across visits, which is exactly what makes it
// safe to collect without asking.

import { sessionId, spanId, traceId } from './ids.js';

const STORAGE_KEY = 'openlog.rum.session';
/** Inactivity after which a new session starts. */
const IDLE_MS = 30 * 60 * 1000;
/** Absolute cap, so a tab left open for days does not report one endless session. */
const MAX_MS = 4 * 60 * 60 * 1000;

interface StoredSession {
  id: string;
  startedAt: number;
  lastSeenAt: number;
  /** Whether this session is sampled in. Decided once, for the whole session. */
  sampled: boolean;
}

/** sessionStorage is unavailable in some privacy modes and inside sandboxed frames; degrade, never throw. */
function readStored(): StoredSession | null {
  try {
    const raw = sessionStorage.getItem(STORAGE_KEY);
    if (!raw) return null;
    const v = JSON.parse(raw) as StoredSession;
    if (typeof v?.id === 'string' && v.id.length === 32 && typeof v.startedAt === 'number') return v;
  } catch {
    /* ignore */
  }
  return null;
}

function writeStored(s: StoredSession): void {
  try {
    sessionStorage.setItem(STORAGE_KEY, JSON.stringify(s));
  } catch {
    /* ignore */
  }
}

export class SessionState {
  private session: StoredSession;
  /** The current page view: its id, and the trace every span of the view belongs to. */
  pageViewId = spanId();
  traceId = traceId();
  /** The page view span's own id, so requests and errors become its children. */
  pageViewSpanId = spanId();

  constructor(private sampleRate: number, now = Date.now()) {
    this.session = this.load(now);
  }

  private load(now: number): StoredSession {
    const stored = readStored();
    if (stored && now - stored.lastSeenAt < IDLE_MS && now - stored.startedAt < MAX_MS) {
      stored.lastSeenAt = now;
      writeStored(stored);
      return stored;
    }
    // Sampling is decided per session, not per event: half a session is not a cheaper session, it is an
    // unreadable one — a page view whose vitals were dropped tells you nothing.
    const fresh: StoredSession = {
      id: sessionId(),
      startedAt: now,
      lastSeenAt: now,
      sampled: Math.random() < this.sampleRate,
    };
    writeStored(fresh);
    return fresh;
  }

  get id(): string {
    return this.session.id;
  }

  get sampled(): boolean {
    return this.session.sampled;
  }

  /** Records activity and rolls the session over when it has gone idle or hit the cap. */
  touch(now = Date.now()): void {
    if (now - this.session.lastSeenAt >= IDLE_MS || now - this.session.startedAt >= MAX_MS) {
      this.session = this.load(now);
      return;
    }
    this.session.lastSeenAt = now;
    writeStored(this.session);
  }

  /** Starts a new page view (a document load or an SPA route change) and returns its ids. */
  newPageView(): { pageViewId: string; traceId: string; spanId: string } {
    this.pageViewId = spanId();
    this.traceId = traceId();
    this.pageViewSpanId = spanId();
    return { pageViewId: this.pageViewId, traceId: this.traceId, spanId: this.pageViewSpanId };
  }

  /** Re-applies a sample rate the server sent; only a session that has not been decided yet can change. */
  applyServerSampleRate(rate: number, now = Date.now()): void {
    if (!(rate > 0 && rate <= 1) || rate === this.sampleRate) return;
    this.sampleRate = rate;
    const stored = readStored();
    // An existing session keeps its decision: flipping mid-visit produces exactly the half-sessions the
    // per-session decision exists to avoid.
    if (!stored) this.session = this.load(now);
  }
}

File: agents/dart/lib/src/session.dart
// The session: a visit, not a person (docs/contracts/mobile-agent.md §4).
//
// 32 random hex characters, expiring after 30 minutes without activity and capped at 4 hours. Both bounds
// matter and they are different questions: the idle timeout ends a visit that stopped, and the cap ends one
// that never stops — a screen left open overnight is not a four-hour visit worth one session id.

import 'ids.dart';

/// How long a session survives without activity.
const Duration sessionIdleTimeout = Duration(minutes: 30);

/// The longest a single session may run, however busy it is.
const Duration sessionMaxAge = Duration(hours: 4);

/// Tracks the current session and the current screen within it.
class SessionState {
  SessionState(this.sampleRate, {DateTime Function()? now}) : _now = now ?? DateTime.now {
    _start(_now());
  }

  final DateTime Function() _now;

  /// The share of sessions kept, decided **once per session** rather than per event: half a session is not
  /// a cheaper session, it is an unreadable one.
  double sampleRate;

  late String id;
  late String traceId;
  late String screenSpanId;
  late bool sampled;
  late DateTime _startedAt;
  late DateTime _lastSeenAt;

  void _start(DateTime at) {
    id = newSessionId();
    traceId = newTraceId();
    screenSpanId = newSpanId();
    _startedAt = at;
    _lastSeenAt = at;
    sampled = _roll();
  }

  bool _roll() {
    if (sampleRate >= 1) return true;
    if (sampleRate <= 0) return false;
    return nextUnit() < sampleRate;
  }

  /// Records activity, starting a new session when the old one has expired. Returns true when a new session
  /// began, which is what a caller needs to know to start a new screen as well.
  bool touch() {
    final at = _now();
    final expired = at.difference(_lastSeenAt) >= sessionIdleTimeout || at.difference(_startedAt) >= sessionMaxAge;
    if (expired) {
      _start(at);
      return true;
    }
    _lastSeenAt = at;
    return false;
  }

  /// A new screen is a new trace: what it causes belongs to it, not to the screen opened minutes ago.
  void newScreen() {
    traceId = newTraceId();
    screenSpanId = newSpanId();
  }

  /// The server applies the sampling weight from the key, so a payload never carries one (§4).
  bool get shouldSend => sampled;
}
